tests/csapi/room_upgrade_test.go

```diff
@@ -0,0 +1,171 @@
+			})
+			// Have alice2 join the room
+			//
+			// We use two users to ensure that all users get taken care of (not just the person
+			// upgrading the room).
+			alice2.MustJoinRoom(t, roomID, nil)
+
+			// Add some push rules
+			alice.MustDo(t, "PUT", []string{"_matrix", "client", "v3", "pushrules", "global", "room", roomID},
+				client.WithJSONBody(t, map[string]interface{}{
+					"actions": []string{"dont_notify"},
+				}),
+			)
+			alice2.MustDo(t, "PUT", []string{"_matrix", "client", "v3", "pushrules", "global", "room", roomID},
+				client.WithJSONBody(t, map[string]interface{}{
+					"actions": []string{"dont_notify"},
+				}),
+			)
+
+			// Sanity check the push rules are in the expected state before the upgrade
+			for _, client := range []*client.CSAPI{alice, alice2} {
+				pushRulesBefore := client.GetAllPushRules(t)
+				must.MatchGJSON(t, pushRulesBefore,
+					match.JSONCheckOff("global.room", []interface{}{
+						roomID,
+					},
+						match.CheckOffMapper(func(r gjson.Result) interface{} { return r.Get("rule_id").Str }),
+						match.CheckOffForEach(func(roomIDFromPushRule interface{}, result gjson.Result) error {
+							return match.JSONKeyEqual("actions.0", "dont_notify")(result)
+						}),
+					),
+				)
+			}
+
+			// Upgrade the room
+			newRoomID := alice.MustUpgradeRoom(t, roomID, "11")
+
+			// Alice2 joins the new room
+			alice2.MustJoinRoom(t, newRoomID, nil)
+
+			// Sanity check the push rules are in the expected state after the upgrade
+			for _, client := range []*client.CSAPI{alice, alice2} {
+				pushRulesAfter := client.GetAllPushRules(t)
+				must.MatchGJSON(t, pushRulesAfter,
+					match.JSONCheckOff("global.room", []interface{}{
+						roomID,
+						newRoomID,
+					},
+						match.CheckOffMapper(func(r gjson.Result) interface{} { return r.Get("rule_id").Str }),
+						match.CheckOffForEach(func(roomIDFromPushRule interface{}, result gjson.Result) error {
+							return match.JSONKeyEqual("actions.0", "dont_notify")(result)
+						}),
+					),
+				)
+			}
+		})
+
+		// When a homeserver becomes aware of a room upgrade (upgrade is done on remote
+		// homeserver), it should copy over any existing push rules for all of its local users
+		// from the old room to the new room at the time of upgrade.
+		t.Run("joining a remote upgraded room carries over existing push rules", func(t *testing.T) {
+			t.Parallel()
+
+			// Alice create a room
+			roomID := alice.MustCreateRoom(t, map[string]interface{}{
+				"preset":       "public_chat",
+				"room_version": "10",
+			})
+			// Remote bob joins the room
+			bob.MustJoinRoom(t, roomID, []spec.ServerName{
+				deployment.GetFullyQualifiedHomeserverName(t, "hs1"),
+			})
+			// Remote bob2 joins the room
+			//
+			// We use two users to ensure that all users get taken care of (not just the first
+			// user on the homeserver).
+			bob2.MustJoinRoom(t, roomID, []spec.ServerName{
+				deployment.GetFullyQualifiedHomeserverName(t, "hs1"),
+			})
```

Comment:
We should sync and wait for the first `bob` to join and `awaitPartialStateJoinCompletion(...)` before joining the second user.

Otherwise, both users could end up racing and using the non-participating code paths ("participating" as in someone on the server is already joined to the room)